### DISCUSSION TITLE
Using middleware causes Links to "external" routes to preload after next v14.

### DISCUSSION THREAD

**Discussion by rjmccluskey** [vercel/next.js#68088] [Help] [answered]
## Description

I have an application that is incrementally adopting Next.js. This means we have our legacy application and a Next.js
application living side by side, with a routing layer above that sends requests to the appropriate application. From
our user's perspective, they see a single application and a single domain.

Inevitably, our Next.js application has links to routes that will ultimately go to the legacy application. We use the
Link component everywhere because as we move pages over to Next.js they will automatically start prefetching when linked
in another Next.js page. Though v13 of next this has worked great, however after upgrading to v14 we noticed an issue.

Now, all those Links with a url pointing to the legacy application will cause invalid requests to _next/data that
respond with 404s.

After much experimentation, I discovered that having middleware defined (no matter how it's used) is the culprit. Even if
I configure the middleware to exclude _next routes, we still get these errors. We do not get the same errors when running
the dev server, only in production.

I've linked a simple reproduction repo. Anyone know if I'm doing something wrong, or is this a bug in Next.js? Are there any workarounds?

## TLDR

If you have middleware (middleware.ts file), the Link component will attempt to preload routes that are not defined in the
Next.js app, resulting in lots of 404s. This only happens in production builds and does not go away even if you configure
the middleware to ignore _next routes.

![Screenshot 2024-07-23 at 4 52 53 PM](https://github.com/user-attachments/assets/104e619e-1b33-4fc3-839b-073df3995868)

### Example

https://github.com/rjmccluskey/next-14-preload-bug

**Comment by rjmccluskey** [comment, marked as answer]
Well after lots of searching I finally came across something! Configuring strict middleware prefetching ended up being the fix. See https://github.com/vercel/next.js/discussions/44596#discussioncomment-4602391.

```js
// next.config.js
{
  experimental: {
      middlewarePrefetch: 'strict',
   },
}
```

**Reply by magnnus** [reply]
Oh! You save me~ Thx~
